docs/howto/advanced/retry.md

```diff
@@ -64,22 +64,44 @@ between retries:
 ## Implementing your own strategy
 
 - If you want to go for a fully fledged custom retry strategy, you can implement your
-  own retry strategy (though we recommend always keeping a max_retry):
+  own retry strategy. This also allows to (optionally) change the priority of the job.
 
+  ```python
+  import random
+  from procrastinate import Job, RetryDecision
+
+  class RandomRetryStrategy(procrastinate.BaseRetryStrategy):
+      max_attempts = 3
+      min = 1
+      max = 10
+
+      def get_retry_decision(self, *, exception:Exception, job:Job) -> RetryDecision:
+          if job.attempts >= max_attempts:
+              return RetryDecision(should_retry=False)
+
+          wait = random.uniform(self.min, self.max)
+
+          return RetryDecision(
+              should_retry=True,
+              schedule_in=wait,
+              new_priority=job.priority + 1,
```

Comment:
Priority is a very specific thing we change here, does it make sense to change other parameters ? If it's really just priority, then fine, otherwise, should this accept `**kwargs` passed to configure the new job ? I think it might make sense to change the queue (e.g you may have task running on small machines and want retries to run on a smaller number of bigger machines)
I don't imagine a usecase for changing queueing lock (I'm not sure I even see when one would use retry and queueing lock)
Changing lock, I don't know, I guess but not sure how to articulate.
Changing schedule in/at: we already do that but what we're doing could be expressed as a special case of changing schedule_at. Also, it might make sense to change decide to retry tomorrow, but in a calendar way and not in 24h (or "retry at night when there's less activity")
Changing the task or the args: really depends on the business case. We could imagine that on certain exception, we'd retry and add a param saying it's safe to skip part of the task.

So all in all: it looks it might make sense to override a large proportion of the job attributes, it seems. What do you think ?

Reply:
(Also: sorry to come up with the idea so late in the dev phase. Had I suggested it earlier, that might have been more respectful of your time, so I want to apologize for this)

Reply:
> Priority is a very specific thing we change here, does it make sense to change other parameters ? If it's really just priority, then fine, otherwise, should this accept `**kwargs` passed to configure the new job ? I think it might make sense to change the queue (e.g you may have task running on small machines and want retries to run on a smaller number of bigger machines) I don't imagine a usecase for changing queueing lock (I'm not sure I even see when one would use retry and queueing lock) Changing lock, I don't know, I guess but not sure how to articulate. 

Yes, sounds like a reasonable suggestion. But I'm not quite convinced to use `**kwargs` yet (I wouldn't even know how to map it to the SQL query). I think I would just add the additional attributes that make sense to the `RetryDecision` (somehow a clearer API, in my opinion). Or what API do you exactly have in mind?

> Changing schedule in/at: we already do that but what we're doing could be expressed as a special case of changing schedule_at. Also, it might make sense to change decide to retry tomorrow, but in a calendar way and not in 24h (or "retry at night when there's less activity") 

Isn't this all possible with `schedule_in`? The library user could take the current date/time and calculate the delta he wants (whether it is in 10 minutes, tomorrow, or ten years). I would keep it simple, but open to other suggestions.

> Changing the task or the args: really depends on the business case. We could imagine that on certain exception, we'd retry and add a param saying it's safe to skip part of the task.

Changing the task sounds like cool and advanced stuff, but on the other hand, it can also be a way to shoot yourself in the foot 😉. When messing up the args, for example, the task would crash instantly. I am also not sure this is really something that library users would do (and I never saw this in a job pipeline before, but that doesn't mean anything). 

And another small detail. I am also not super convinced of the `should_retry` attribute. Why not just return None when no retry should take place?

> (Also: sorry to come up with the idea so late in the dev phase. Had I suggested it earlier, that might have been more respectful of your time, so I want to apologize for this)

No worries. Sometimes, you have to have things in front of you to come up with new ideas.

Reply:
Agree with all your points :)

Reply:
How about ...
```python
@attr.dataclass
class RetryDecision:
    schedule_in: float | None = None
    priority: int | None = None
    queue: str | None = None
    lock: str | None = None
```
I would only use the `new_` prefix  (for example `new_priority`) in the SQL query or do you like that naming in the `RetryDecision` with the `new_` prefix?
Regarding changing the `args`, I would prefer to add something like a `meta` field where additional information for that job could be stored. But I wouldn't do that on this issue; I would do it more when there is really a need for that.